synapse/handlers/room.py

```diff
@@ -1220,6 +1221,114 @@ async def filter_evts(events: List[EventBase]) -> List[EventBase]:
+            room_id: Room to fetch the event from
+            timestamp: The point in time we should navigate from in the given
+                direction to find the closest event.
+            direction: ["f"|"b"] to indicate whether we should navigate forward
+                or backward from the given timestamp to find the closest event.
+
+        Returns:
+            `event_id` closest to the given timestamp in the given direction
+
+        Raises:
+            SynapseError if unable to find any event locally in the given direction
+        """
+
+        local_event_id = await self.store.get_event_for_timestamp(
+            room_id, timestamp, direction
+        )
+
+        logger.debug(
+            "get_event_for_timestamp: locally, we found event_id=%s closest to timestamp=%s",
+            local_event_id,
+            timestamp,
+        )
+
+        # If we found a gap, we should probably ask another homeserver first
+        # about more history in between
+        if local_event_id:
+            event_next_to_gap = await self.store.event_next_to_gap(
+                room_id, local_event_id
+            )
+
+        if not local_event_id or event_next_to_gap:
+            logger.debug(
+                "get_event_for_timestamp: locally, we found event_id=%s closest to timestamp=%s which is next to a gap in event history so we're asking other homeservers first",
+                local_event_id,
+                timestamp,
+            )
+
+            # Find other homeservers from the given state in the room
+            curr_state = await self.state_handler.get_current_state(room_id)
+            curr_domains = get_domains_from_state(curr_state)
+            likely_domains = [
+                domain for domain, depth in curr_domains if domain != self.server_name
+            ]
+
+            # Loop through each homeserver candidate until we get a succesful response
+            for domain in likely_domains:
+                try:
+                    remote_response = await self.transport_layer.timestamp_to_event(
+                        domain, room_id, timestamp, direction
+                    )
+                    logger.debug(
+                        "get_event_for_timestamp: response from domain(%s)=%s",
+                        domain,
+                        remote_response,
+                    )
+                    if not remote_response:
+                        continue
+
+                    remote_event_id = remote_response.get("event_id", None)
+                    if remote_event_id:
+                        # TODO: Do we want to persist this as an extremity?
+                        # TODO: I think ideally, we would try to backfill from
+                        # this event and run this whole
+                        # `get_event_for_timestamp` function again to make sure
+                        # they didn't give us an event from their gappy history.
+                        return remote_event_id
```

Comment:
I wonder if it'd be a good idea to amend the MSC to include `origin_server_ts` in the response, so we could discard events that are worse (further away) than the one we found locally.

Reply:
Seems decent 👍 Everything is up in the air and we can change anything. I am even considering whether we can add a pagination token to start paginating on `/messages` right away, https://github.com/matrix-org/synapse/pull/9445#discussion_r747943150

Reply:
Added `origin_server_ts` to the federation endpoint and the optimization to only return the remote event if closer than the local event.

Is it worth adding to the client endpoint as well?

Reply:
I've added it to the client endpoint as well because it could be useful to make sure that a requested event is not 24 hours older than the timestamp requested for day-by-day archive for example.